### PR TITLE
HPCC-17555 Fix various chained stop() issues and conditional loop initialization

### PR DESCRIPTION
**Pull request by jakesmith**
A number of activities would not call their input stop() methods
if they had never been started.
Also, a loop graph would not initialize all activities if some
activities were conditional branches.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:

Specific modified test (from PR of HPCC-17179) ran / passed (looprpn2.ecl)
Full regression suite ran/passed.

### CODE REVIEW

**Comment by jakesmith** [comment]
@ghalliday - please review.

I was dubious whether this should be targeted at 6.4.0, as late/largish change.

**Comment by richardkchapman** [comment]
6.4.0 testing has not really started in earnest... if it's a bugfix I'm ok with it going in 6.4.0

**Comment by richardkchapman** [comment]
@ghalliday Please review

File: thorlcr/activities/indexwrite/thindexwriteslave.cpp [hunk 1]
```diff
@@ -537,7 +528,8 @@ class IndexWriteSlaveActivity : public ProcessSlaveActivity, public ILookAheadSt
     {
         if (processed & THORDATALINK_STARTED)
         {
-            doStopInput();
+            if (!inputStopped) // probably already stopped in process()
```
Comment (ghalliday) [inline thorlcr/activities/indexwrite/thindexwriteslave.cpp]: minor: is there any advantage making this conditional?

**Review by ghalliday — COMMENTED** [review]
Code looks cleaner, and even better code size went down.

File: thorlcr/activities/indexwrite/thindexwriteslave.cpp (hunk 1 above)
Comment (jakesmith) [inline thorlcr/activities/indexwrite/thindexwriteslave.cpp]: there really shouldn't be. But I kept it conditional as before to minimize this changes impact.

**Pull request merged**